### PR TITLE
[Reasoning][Feature] Support for speculative decoding with thinking budget

### PR DESCRIPTION
**Pull request by rishitdholakia13**
This PR provide support and and compatibility for thinking budget with speculative decoding.
For further details on implementation please refer : https://github.com/vllm-project/vllm/pull/34668#issuecomment-4066967952

The PR extends to refactor the thinking budget design by moving it out of the logitsprocessor. 
This PR provides the following compatibilities with thinking budget:
- [x] Speculative decoding
- [x] - Non speculative decoding
- [x] - Structured output  + Speculative decoding
- [x] - Async
- [x] - Non Async
- [x] CI test for speculative decoding + thinking budget in a concurrent request setting with a mix of requests containing both thinking budget and no thinking budget

### CODE REVIEW

**Comment by rishitdholakia13** [comment]
Thank you for the suggestion @njhill. , I have  the following pseudo code, that I can use to get it work without the logitsprocessor and similar to the way penalties are being worked with. In that is the case, can we get the #20859  merged. Since that uses the changes required for the thinking budget interface and some of the code thinking budget logic. I am developing on top of the #20859 PR. In the meantime, can I get some suggestions and review on the refactored design without the logitsprocessor. 



# Thinking Budget Refactor — Design Flow

Move thinking budget out of the logits processor into a **ThinkingBudgetStateHolder**. Sampler/RejectionSampler call **update_state_and_apply** after combining outputs with spec tokens (same pattern as penalties).

---

## 1. ThinkingBudgetStateHolder

| Method | When | What |
|--------|------|------|
| `sync_batch(batch_update)` | `refresh_metadata` | Add/remove/move state only. **No** `_update_think_state`. |
| `update_state_and_apply(...)` | Sampler / RejectionSampler | Feed token history → `_update_think_state` → apply to logits. **Only** place _update_think_state runs. |

```python
def sync_batch(self, batch_update):
    for i in batch_update.removed: self._state.pop(i, None)
    for i, params, prompt, out in batch_update.added:
        if params.thinking_token_budget is not None:
            self._state[i] = self._init_state_entry(prompt, params.thinking_token_budget)
    for i1, i2, direction in batch_update.moved: ...

def update_state_and_apply(self, logits, output_token_ids, spec_token_ids, ...):
    for req_idx, state in self._state.items():
        state["output_tok_ids"] = output_token_ids[req_idx]
        self._update_think_state(state)
    # then: build mask/force from state, logits[active, force_ids] = 1e9
```

---

## 2. Sampler

Combine with spec when thinking (or penalties); after penalties, call holder.

```python
if predict_bonus_token and (not no_penalties or not no_thinking_budget or bad_words):
    output_token_ids = self._combine_outputs_with_spec_tokens(output_token_ids, spec_token_ids)
logits = self.apply_penalties(...)
if not no_thinking_budget and thinking_budget_state_holder:
    thinking_budget_state_holder.update_state_and_apply(logits, output_token_ids ..., predict_bonus_token=True)
```

---

## 3. RejectionSampler

Same: combine (expanded list), build `repeat_indices`, then apply with `repeat_indices` / `num_draft_tokens`. We do not update the state here we have already updated the state in sampler as the bonus token's logits processor is called before the target logits's logitsprocessor. The sampler also handles the state for the draft tokens. 

```python
output_token_ids = self._combine_outputs_with_spec_tokens(...)  # expanded
repeat_indices = torch.arange(num_requests).repeat_interleave(num_draft_tokens)
# ... penalties, logitsprocs ...
if not no_thinking_budget and thinking_budget_state_holder:
    thinking_budget_state_holder.apply(..., repeat_indices=repeat_indices, num_draft_tokens=metadata.num_draft_tokens) 
```

---

## 4. GPUInputBatch

Create holder when thinking enabled; sync_batch in refresh_metadata; pass holder + static params in metadata.

```python
# __init__: self.thinking_budget_state_holder = ThinkingBudgetStateHolder(...) if thinking_enabled else None
def refresh_metadata(self):
    ...
    if self.thinking_budget_state_holder:
        self.thinking_budget_state_holder.sync_batch(batch_update)
# _make_sampling_metadata: pass no_thinking_budget, thinking_token_budgets, think_start/end_token_ids, thinking_budget_state_holder
```

---

## 5. Cleanup

Remove `ThinkingTokenBudgetLogitsProcessor` from builtin and from `BUILTIN_LOGITS_PROCESSORS` / spec-decode branch.

---

## Flow summary

```
refresh_metadata  →  state_holder.sync_batch(add/remove/move)
_make_sampling_metadata  →  pass holder + static params
Sampler / RejectionSampler  →  _combine_outputs_with_spec_tokens  →  apply_penalties  →  state_holder.update_state_and_apply
```

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -1119,6 +1115,19 @@ def _update_states(self, scheduler_output: "SchedulerOutput") -> None:
...
+
+        # Update spec_token_ids with draft_tokens from the last step
+        # if async scheduling and in spec mode
+        # DO the update here, becaus it is needed for thinking budget 
+        # in speculative decoding mode.
+        if self.use_async_scheduling:
+            if self._draft_token_req_ids is not None:
+                draft_token_ids_cpu, _ = self._get_draft_token_ids_cpu()
+                self.input_batch.update_async_spec_token_ids(draft_token_ids_cpu)
+            self.input_batch.update_async_output_token_ids()
```
Comment (njhill) [inline vllm/v1/worker/gpu_model_runner.py]: I think moving these syncs to here (before the forward pass) may negate much of the performance benefit of async scheduling.

Comment (rishitdholakia13) [reply vllm/v1/worker/gpu_model_runner.py]: With this change—moving the logic out of the logits processor—we eliminate the need for this modification altogether, so it is no longer a concern.

File: =4.5.1
Comment (njhill) [inline =4.5.1:1]: Looks like this file was added by accident?

File: vllm/v1/sample/metadata.py [hunk 2]
```diff
@@ -47,3 +50,5 @@ class SamplingMetadata:
 
     # Speculative token ids
     spec_token_ids: list[list[int]] | None = None
+    thinking_budget_state_holder: ThinkingBudgetStateHolder | None = None
+    no_thinking_budget: bool = True
```
Comment (njhill) [inline vllm/v1/sample/metadata.py]: Is this needed? Can this just be indicated by `thinking_budget_state_holder is None`?

File: vllm/v1/worker/gpu_model_runner.py [hunk 3]
```diff
@@ -613,6 +616,9 @@ def __init__(
         )
         self._init_block_sizes = [placeholder_block_size]
         self._init_kernel_block_sizes = [placeholder_block_size]
+        self.thinking_budget_state_holder = maybe_create_thinking_budget_state_holder(
+            self.vllm_config, self.device, self.pin_memory
+        )
```
Comment (njhill) [inline vllm/v1/worker/gpu_model_runner.py]: Is this needed here? could it be moved into the `InputBatch` constructor?

**Review by njhill — COMMENTED** [review]
Thanks @rishitdholakia13, this looks good to me.

I haven't reviewed all of the new logic in detail but in terms of how it's integrated into the existing code it looks ok.

File: =4.5.1
Comment (rishitdholakia13) [reply =4.5.1:1]: Yes, will remove this

File: vllm/v1/sample/metadata.py (hunk 2 above)
Comment (rishitdholakia13) [reply vllm/v1/sample/metadata.py:54]: Yes, I have looked into refactoring the code, accordingly

File: vllm/v1/worker/gpu_model_runner.py (hunk 3 above)
Comment (rishitdholakia13) [reply vllm/v1/worker/gpu_model_runner.py:624]: These were some of the reasons why I had added it model runner : 

- InputBatch is not always the same object — after KV cache setup the runner may construct a new InputBatch when block sizes differ from the placeholder; anything created only inside InputBatch.__init__ would be recreated and lose state.

- ThinkingBudgetStateHolder must stay one long-lived instance — it holds per-request and GPU sampling state that must survive that swap, so we create it once (on the runner) and pass the same reference into each new InputBatch.

- `maybe_create_thinking_budget_state_holder` needs VllmConfig (reasoning settings, max_num_seqs, speculative decode flags, device, pin memory). InputBatch today is built from mostly derived scalars, not the full config. Moving creation into InputBatch means either threading vllm_config (or a similar bundle) into InputBatch, which widens its dependencies, or duplicating those fields at the call site—so keeping creation next to the runner (which already owns vllm_config) avoids that API creep unless you explicitly want it.

**Comment by rishitdholakia13** [comment]
Hey @njhill , I’ve addressed the feedback from the initial review—could you take another look ? It seems there’s strong community interest in getting this feature merged soon. #39573.

I have added additional e2e tests  to tighten the right functioning of the thinking budget logic with and with out SD.

File: vllm/v1/worker/gpu_model_runner.py (hunk 3 above)
Comment (njhill) [reply vllm/v1/worker/gpu_model_runner.py]: @rishitdholakia13 I don't think the first two points matter. There's only one InputBatch, if it is recreated that happens during startup so it shouldn't matter if the TB state is also recreated.

Re the third point, most of these things are already passed to the InputBatch constructor. I think the only additional things you'd need to add would be:
- num_spec_tokens (0 means no spec decoding) .. actually this could replace the existing `is_spec_decode` arg
- reasoning_config .. could replace thinking_budget_state_holder

**Review by njhill — COMMENTED** [review]
Thanks @rishitdholakia13. Just had one follow-on comment.

**Comment by njhill** [comment]
@rishitdholakia13 also please fix the pre-commit errors!

File: vllm/v1/worker/gpu_input_batch.py [hunk 4]
```diff
@@ -92,12 +96,21 @@ def __init__(
...
+            reasoning_config,
+            max_num_reqs,
+            num_spec_tokens,
+            device,
+            pin_memory,
+        )
+        self.thinking_token_budget_reqs: set[str] = set()
         self.is_pooling_model = is_pooling_model
-        self.is_spec_decode = is_spec_decode
+        self.num_spec_tokens = num_spec_tokens
```
Comment (njhill) [inline vllm/v1/worker/gpu_input_batch.py]: nit: it doesn't look like this is actually used (the field it replaces wasn't either), so could be removed I think

**Review by njhill — APPROVED** [review]
Thanks @rishitdholakia13 LGTM, just a small comment

File: vllm/v1/worker/gpu_input_batch.py (hunk 4 above)
Comment (rishitdholakia13) [reply vllm/v1/worker/gpu_input_batch.py]: Yes,  I noticed it as well, thank you for pointing it out. Will remove it.

**Comment by rishitdholakia13** [comment]
Hey @njhill , thank you for approving the PR. I see that the CI does not seem to be running. I guess a ready label needs to be added to run the CI. I would want to make sure the CI for e2e thinking budget test passes. Thank you
